src/pybind/mgr/volumes/fs/subvolume.py

```diff
@@ -252,56 +202,63 @@ def _get_ancestor_xattr(self, path, attr):
-
-    def get_subvolume_path(self, subvolume_path):
-        path = self._subvolume_path(subvolume_path)
-        try:
-            self.fs.stat(path)
-        except cephfs.ObjectNotFound:
-            return None
-        return path
-
-    def _snapshot_path(self, dir_path, snapshot_name):
-        return os.path.join(
-            dir_path, self.rados.conf_get('client_snapdir'), snapshot_name
-        )
-
-    def _snapshot_create(self, dir_path, snapshot_name, mode=0o755):
+    def _snapshot_create(self, snappath, mode=0o755):
         """
         Create a snapshot, or do nothing if it already exists.
         """
-        snapshot_path = self._snapshot_path(dir_path, snapshot_name)
         try:
-            self.fs.stat(snapshot_path)
+            self.fs.stat(snappath)
         except cephfs.ObjectNotFound:
-            self.fs.mkdir(snapshot_path, mode)
+            self.fs.mkdir(snappath, mode)
         else:
-            log.warn("Snapshot '{0}' already exists".format(snapshot_name))
-
+            log.warn("Snapshot '{0}' already exists".format(snappath))
 
-    def _snapshot_delete(self, dir_path, snapshot_name):
+    def _snapshot_delete(self, snappath):
         """
         Remove a snapshot, or do nothing if it doesn't exist.
         """
-        snapshot_path = self._snapshot_path(dir_path, snapshot_name)
-        self.fs.stat(snapshot_path)
-        self.fs.rmdir(snapshot_path)
+        self.fs.stat(snappath)
+        self.fs.rmdir(snappath)
 
-    def create_subvolume_snapshot(self, subvolume_path, snapshot_name, mode=0o755):
-        return self._snapshot_create(self._subvolume_path(subvolume_path), snapshot_name, mode)
+    def create_subvolume_snapshot(self, spec, snapname, mode=0o755):
+        snappath = spec.make_subvol_snap_path(self.rados.conf_get('client_snapdir'), snapname)
+        self._snapshot_create(snappath, mode)
 
-    def delete_subvolume_snapshot(self, subvolume_path, snapshot_name):
-        return self._snapshot_delete(self._subvolume_path(subvolume_path), snapshot_name)
+    def remove_subvolume_snapshot(self, spec, snapname):
+        snappath = spec.make_subvol_snap_path(self.rados.conf_get('client_snapdir'), snapname)
+        self._snapshot_delete(snappath)
 
-    def create_group_snapshot(self, group_id, snapshot_name, mode=0o755):
-        return self._snapshot_create(self._group_path(group_id), snapshot_name, mode)
+    def create_group_snapshot(self, spec, snapname, mode=0o755):
+        snappath = spec.make_group_snap_path(self.rados.conf_get('client_snapdir'), snapname)
+        self._snapshot_create(snappath, mode)
 
-    def delete_group_snapshot(self, group_id, snapshot_name):
-        return self._snapshot_delete(self._group_path(group_id), snapshot_name)
+    def remove_group_snapshot(self, spec, snapname):
+        snappath = spec.make_group_snap_path(self.rados.conf_get('client_snapdir'), snapname)
+        return self._snapshot_delete(snappath)
+
+    def connect(self):
```

Comment:
Not sure why you shuffled connect() and the below methods to the end of the module?

Reply:
the internal context manager routine overrides (and helper routines) were right in between the subvolume logic -- https://github.com/ceph/ceph/blob/master/src/pybind/mgr/volumes/fs/subvolume.py#L92 (between `_group_path()` and `_mkdir_p()`).

Reply:
OK